caffe2/opt/converter.cc

```diff
@@ -295,15 +297,26 @@ repr::NNModule convertToNNModule(caffe2::NetDef &net, std::unordered_map<std::st
   auto bbNode =
       cfg.createNode(util::make_unique<repr::BasicBlockType<repr::NNGraph>>());
 
+  std::unordered_map<std::string, repr::NNGraph::NodeRef> entryNodeMap;
+  std::unordered_map<std::string, repr::NNGraph::NodeRef> exitNodeMap;
+
+  for (auto external_input : net.external_input()) {
+    entryNodeMap[external_input] = nullptr;
+  }
+  for (auto external_output : net.external_output()) {
+    exitNodeMap[external_output] = nullptr;
+  }
+
   for (auto &op : *net.mutable_op()) {
     auto opNode = dfg.createNode(); // Create an empty node for the operator.
     // First calculate in-edges (data dependencies).
     for (const auto &input : op.input()) {
       // If we've never seen this tensor, make one.
       if (!blobMap.count(input)) {
-        auto tensor = util::make_unique<repr::Tensor>(input);
-        blobMap[input] =
-            dfg.createNode(unique_dyn_cast<repr::NeuralNetData>(tensor));
+        blobMap[input] = dfg.createNode(util::make_unique<repr::Tensor>(input));
+        if (entryNodeMap.count(input) && !entryNodeMap[input]) {
```

Comment:
This one might be helpful? 
https://github.com/pytorch/pytorch/blob/master/caffe2/utils/map_utils.h#L13

Reply:
not sure what you mean by this?

the goal of this line is to take an `entryNodeMap` populated with nullptrs and instead populate it with real pointers.  the reason for this is to efficiently check if the input blob name exists within the `external_inputs` field of the netdef (because querying a map is faster than a protobuf repeated field)

Reply:
I mean you can replace this condition with 
```
if(!get_default(entryNodeMap, input, nullptr))
```
Just better code reuse and it saves you one hash lookup. Logic wise the code is correct.

Reply:
in fact this is done with 3 lookups right now (the whole if statement). It can be 1 find() call. Then then you can update the value. Also value doesn't have to be nullptr checked, I think. As blobMap[input] should  not change as well. 

Also therei s no reason to do blobMap[input] as you already know the value. So total we have 4 lookups here which can be replaced with 1 :)